crates/wiggle/src/lib.rs

```diff
@@ -582,19 +597,79 @@ impl<'a, T> GuestPtr<'a, [T]> {
         })?;
 
         // Validate all elements in slice.
-        // SAFETY: ptr has been validated by self.mem.validate_size_align
+        // SAFETY: `ptr` has been validated by `self.mem.validate_size_align`
         for offs in 0..self.pointer.1 {
             T::validate(unsafe { ptr.add(offs as usize) })?;
         }
 
-        // SAFETY: iff there are no overlapping borrows it is valid to construct a &mut [T]
+        // SAFETY: iff there are no overlapping borrows it is valid to construct
+        // a `&mut [T]`.
         let ptr = unsafe { slice::from_raw_parts_mut(ptr, self.pointer.1 as usize) };
 
-        Ok(GuestSliceMut {
+        Ok(Some(GuestSliceMut {
             ptr,
             mem: self.mem,
             borrow,
-        })
+        }))
+    }
+
+    /// Similar to `as_slice_mut`, this function will attempt to create a smart
+    /// pointer to the WebAssembly linear memory. All validation and Wiggle
+    /// borrow checking is the same, but unlike `as_slice_mut`, the returned
+    /// `&mut` slice can point to WebAssembly shared memory. Though the Wiggle
+    /// borrow checker can guarantee no other Wiggle calls will access this
+    /// slice, it cannot guarantee that another thread is not modifying the
+    /// `&mut` slice in some other way. Thus, access to that slice is marked
+    /// `unsafe`.
+    pub fn as_unsafe_slice_mut(&self) -> Result<UnsafeGuestSliceMut<'a, T>, GuestError>
+    where
+        T: GuestTypeTransparent<'a>,
+    {
+        // SAFETY: see `as_slice_mut`.
+        let len = self.checked_byte_len()?;
+        let ptr =
+            self.mem
+                .validate_size_align(self.pointer.0, T::guest_align(), len)? as *mut T;
+
+        let borrow = self.mem.mut_borrow(Region {
+            start: self.pointer.0,
+            len,
+        })?;
+        for offs in 0..self.pointer.1 {
+            T::validate(unsafe { ptr.add(offs as usize) })?;
+        }
+        let ptr = unsafe { slice::from_raw_parts_mut(ptr, self.pointer.1 as usize) };
+
+        // Wrap up a `GuestSliceMut` in its `unsafe` container.
+        let guest_slice_mut = GuestSliceMut {
+            ptr,
+            mem: self.mem,
+            borrow,
+        };
+        Ok(UnsafeGuestSliceMut(guest_slice_mut))
+    }
```

Comment:
Could this perhaps be refactored with the result or such to be usable as an internal implementation detail of the two methods above? Deduplicating the bounds checks would be pretty convenient and the "Unsafe" version could then have a methods `{shared,mut}_borrow` to give the safe versions, returning `None` for shared memories.

basically `as_slice()` could be a small wrapper around `Ok(self.as_unsafe_slice().shared_borrow())` or something like that.